custom_metrics/almanac.js

```diff
@@ -7,37 +7,223 @@
 // 3. Test your change by following the instructions at https://github.com/HTTPArchive/almanac.httparchive.org/issues/33#issuecomment-502288773.
 // 4. Submit a PR to update this file.
 
-return {
+function parseNodes(nodes) {
+  var parsedNodes = [];
+  if (nodes) {
+    for (var i = 0, len = nodes.length; i < len; i++) {
+      var node = nodes[i];
+      var attributes = Object.values(node.attributes);
+      var el = {};
+
+      el.tagName = node.tagName.toLowerCase(); // for reference
+      for (var n = 0, len2 = attributes.length; n < len2; n++) {
+        var attribute = attributes[n];
+        el[attribute.name.toLowerCase()] = attribute.value;
+      }
+
+      parsedNodes.push(el);
+    }
+  }
+  return parsedNodes;
+}
+
+return JSON.stringify({
   // Wether the page contains <script type=module>.
   '01.12': document.querySelector('script[type=module]') ? 1 : 0,
   // Wether the page contains <script nomodule>.
   '01.13': document.querySelector('script[nomodule]') ? 1 : 0,
   'link-nodes': (() => {
     // Returns a JSON array of link nodes and their key/value attributes.
-    // Used by 01.14, 01.15, 01.16...
+    // Used by 01.14, 01.15, 01.16, 10.6,  06.46, 12.18
     var nodes = document.querySelectorAll('head link');
-    var linkNodes = [];
+    var linkNodes = parseNodes(nodes);
+
+    return linkNodes;
+  })(),
+  'meta-nodes': (() => {
+    // Returns a JSON array of meta nodes and their key/value attributes.
+    // Used by 10.6, 10.7 (potential: 09.29, 12.5, 04.5)
+    var nodes = document.querySelectorAll('head meta');
+    var metaNodes = parseNodes(nodes);
+
+    return metaNodes;
+  })(),
+  // Extract schema.org elements and finds all @context and @type usage
+  '10.5': (() => {
+    function nestedLookup(items) {
+      var keys = Object.keys(items);
+      for (var i = 0, len = keys.length; i < len; i++) {
+        var item = items[keys[i]];
+        // if array or object, dive into it
+        if (item instanceof Object || item instanceof Array) {
+          nestedLookup(item);
+        }
+      }
+      if (items['@type']) {
+        if (items['@context']) {
+          link.href = items['@context'] + '/' + items['@type'];
+          schemaElements[link.hostname + link.pathname] = true;
+        } else {
+          schemaElements[items['@type']] = true;
+        }
+      }
+    }
+
+    var nodes = document.querySelectorAll('[itemtype], script[type=\'application/ld+json\']');
+    var link = document.createElement('a');
+    var schemaElements = {};
+
+    if (nodes) {
+      for (var i = 0, len = nodes.length; i < len; i++) {
+        var node = nodes[i];
+        var item = node.getAttribute('itemtype');
+
+        if (item) {
+          // microdata
+          link.href = node.getAttribute('itemtype');
+          schemaElements[link.hostname + link.pathname] = true;
+        } else if ((node.tagName = 'SCRIPT')) {
+          // json+ld
+          var content = JSON.parse(node.textContent);
```

Comment:
We need to ensure that the entire custom metric doesn't fail if the JSON-LD is unable to be parsed. And in general, we should be very careful about building in resiliency because web pages can be... weird.

Can you add a try/catch around this? Any other places that you can think of that could better handle failures?

Reply:
Good catch! Added it. As well for a querySelector based on _user_ input.